#411 Data processing and extraction nodes and script docs (Houdini) (#413)
Inbibo/adonisfx_docs: docs/houdini/scripts/data_extraction.md

The data extraction script allows users to:

- Extract joint transform input data from a KineFX joint source.
- Extract mesh displacement output data from a rest skin and a simulated skin.
- Optionally extract muscle activation data from Adonis muscle solvers.
- Record data over the full playback range or over specific frame windows.
- Skip frames to reduce redundant poses in the extracted dataset.
- Recook frames for stabilization before recording data.
- Export the resulting dataset files to a target directory.

The extraction script is the Python API equivalent of the workflow exposed through the [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool). In most production workflows, the data should first be prepared with the [AdnMLDataProcessing HDA](../tools/ml_data_processing), and then extracted either through the TOP HDA or by calling this script directly.

The extraction process writes the generated dataset into the target folder specified by `save_directory_path`. The main exported files are `inputs.csv`, `outputs.csv`, `joints.json`, and `extraction_config.json`.

> [!NOTE]

> > Data extraction for ML training is only supported with the **AdonisML** bundle. Make sure the Adonis plugin is installed and licensed to support ML before running this workflow.

Reviewer: Can we convert this into an IMPORTANT note with this sentence?

```
> [!IMPORTANT]
> An Adonis ML license is required to use this feature.
```

Please, move it to the very top of the page.
Also, please do the same in the other pages.

Author: Done, will port this over to the other PR's

> > [!WARNING]

Reviewer: I would not use Warning here. To me these are requirements and I would just list them in plain text as in other pages

Author: Done

## Extract Data

To extract ML training data from Python, run this command in Houdini:

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">from adnml.scripts.houdini import data_extraction

data_extraction.extract(
    rest_skin_path="/obj/geo1/REST_SKIN",
    sim_skin_path="/obj/geo1/OUT_SKIN",
    joints_path="/obj/geo1/OUT_JOINTS",
    joint_names_list=["root", "spine", "neck", "head"],
    save_directory_path="path/to/export/folder"
)
</code></pre>

The extraction process writes the dataset files to the folder specified by `save_directory_path`.

The required arguments are:

- `rest_skin_path`: Path to the rest skin SOP node.
- `sim_skin_path`: Path to the simulated skin SOP node.
- `joints_path`: Path to the joint source SOP node.
- `joint_names_list`: List of Houdini/KineFX joint names to extract.
- `save_directory_path`: Target folder where the extracted dataset will be saved.

> > [!NOTE]

Reviewer: Are we abusing of NOTEs as well? I am just thinking that NOTEs are to add some auxiliary info, not part of the main explanation. Just a thought! You can leave it

Author: Done

The optional arguments are:

- `muscles_paths`: Muscle SOP paths used to extract muscle activation data. This can be a single string, a list of strings, or a UI-style string separated by spaces, commas, tabs, or newlines. The recommended setup is to point to a merge node that combines all **ADN_OUT_** Null nodes coming from the muscle nodes. This is supported and keeps the extraction input centralized. The paths can also point to individual **AdnMuscle** SOPs, muscle geometry SOPs, or **ADN_OUT_** nodes.
- `skip_frames`: Number of frames to skip between recorded poses. This helps reduce redundant pose data and extract more diverse samples.
- `stabilization_frames`: Number of times to recook a frame before recording displacement data. This parameter damps the motion inertia in the recorded poses.
- `frame_windows`: List of frame windows to record. If empty, the entire playback range is recorded.
- `force_overwrite`: If enabled, existing dataset files in the target folder can be overwritten.

## Frame Windows

By default, the extraction process uses the current Houdini playback range.

To extract only specific parts of the animation, provide the `frame_windows` argument:

> <pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">frame_windows = [[1, 120], [220, 360], [500, 720]]

Reviewer: I would remove this, the example with the full call is descriptive enough in my opinion

Author: Done

</code></pre>

Each entry defines a frame range using the format `[start_frame, end_frame]`.

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">from adnml.scripts.houdini import data_extraction

data_extraction.extract(
    rest_skin_path="/obj/geo1/REST_SKIN",
    sim_skin_path="/obj/geo1/OUT_SKIN",
    joints_path="/obj/geo1/OUT_JOINTS",
    joint_names_list=["root", "spine", "neck", "head"],
    save_directory_path="path/to/export/folder",
    frame_windows=[[1, 120], [220, 360], [500, 720]]
)
</code></pre>

Frame windows are useful when only specific animation ranges should contribute to the training dataset.

## Skip Frames

The `skip_frames` argument controls how many frames are skipped between recorded poses.

> <pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">skip_frames = 2

Reviewer: I would remove this

Author: Done

</code></pre>

This can be used to reduce redundant pose data and extract more diverse samples.

Lower values are recommended for fast animations, while higher values can be used for slower animations. Typical suggested values for normal animation speeds are between `2` and `5`.

The skip frames will be computed from the start of each frame window, this ensures that the starting frame of each window is always recorded in the dataset.

## Stabilization Frames

The `stabilization_frames` argument controls how many times each recorded frame should be recooked before displacement data is computed and written.

> <pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">stabilization_frames = 5

Reviewer: Same here, I would remove this

Author: Done

- A single SOP path pointing to one muscle output.
- A list of SOP paths.
- A string containing multiple paths separated by spaces, commas, tabs, or newlines.

The paths can point to **AdnMuscle** SOPs, muscle geometry SOPs, or **ADN_OUT_** nodes.

When muscle paths are provided, the extracted data will support training the ML model on material properties prediction for AdnSmartTissue. If this input is not provided, the extracted data will support training models for AdnMLDeformer only.

## Force Overwrite

By default, the extraction process will not overwrite existing dataset files.

If `inputs.csv` or `outputs.csv` already exist in the target export folder, the extraction will raise an error unless `force_overwrite` is enabled.

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">from adnml.scripts.houdini import data_extraction

data_extraction.extract(
    rest_skin_path="/obj/geo1/REST_SKIN",
    sim_skin_path="/obj/geo1/OUT_SKIN",
    joints_path="/obj/geo1/OUT_JOINTS",
    joint_names_list=["root", "spine", "neck", "head"],
    save_directory_path="path/to/export/folder",
    force_overwrite=True
)
</code></pre>

> [!WARNING]

> > Enabling `force_overwrite` will replace any existing extraction data in the target folder, use it with caution.

Reviewer: perfect use case for Warning messages!

Author: Done

- The simulation skin path is valid.
- The joint source path is valid.
- The save directory path is valid.
- The rest skin and simulation skin are not the same object.
- The rest skin and simulation skin have matching topology.
- The rest skin and simulation skin contain point positions.
- The joint source geometry contains valid KineFX joint transform data.
- Every requested joint name exists in the joint source geometry.
- The skip frame value is a valid integer greater than or equal to `0`.
- The stabilization frame value is a valid integer greater than or equal to `0`.
- The frame windows are valid.
- Existing dataset files are not overwritten unless `force_overwrite` is enabled.

## Errors

The `extract` function may raise the following errors:

- `ValueError`: Raised when an input path, frame setting, geometry, or joint list is invalid.
- `FileExistsError`: Raised when dataset files already exist and `force_overwrite` is disabled.

## Limitations

- The rest skin and simulation skin must have the same topology.
- The joint source geometry must contain the requested KineFX joint names.
- The joint source geometry must contain valid KineFX joint transform data.

> - Existing `inputs.csv` or `outputs.csv` files will prevent extraction unless `force_overwrite` is enabled.

Reviewer: I have the same feeling that in other pages, this list sounds more like a list of requirements than limitations

Author: Done

# Data Extraction

In the `adnml.scripts.houdini.data_extraction` module, a Python function is provided to extract machine learning training data from a processed Adonis simulation setup.

The data extraction script allows users to:

- Extract joint transform input data from a KineFX joint source.
- Extract mesh displacement output data from a rest skin and a simulated skin.

> - Optionally extract muscle activation data from Adonis muscle geometries.

Reviewer 2: Suggestion:
```suggestion
- Optionally extract muscle activation data from Adonis muscle solvers.
```

Author: Done, good catch

- Record data over the full playback range or over specific frame windows.
- Skip frames to reduce redundant poses in the extracted dataset.
- Recook frames for stabilization before recording data.
- Export the resulting dataset files to a target directory.

The extraction script is the Python API equivalent of the workflow exposed through the [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool). In most production workflows, the data should first be prepared with the [AdnMLDataProcessing HDA](../tools/ml_data_processing), and then extracted either through the TOP HDA or by calling this script directly.

The extraction process writes the generated dataset into the target folder specified by `save_directory_path`. The main exported files are `inputs.csv`, `outputs.csv`, `joints.json`, and `extraction_config.json`.

> [!WARNING]
> Before running data extraction, an AdnMLDataProcessing HDA must be configured in the Houdini scene. The `sim_skin_path` and `joints_path` arguments must point to the outputs generated by this processing HDA. The simulated skin should point to the **ADN_OUT_SKIN** output and the joints should point to the **ADN_OUT_ML_JOINTS** output.
>

> > If the scene has not been prepared with an AdnMLDataProcessing HDA, the extraction will fail because the script expects the simulation skin and joint source to contain the data generated by the processing step. Advanced users may automate this preparation themselves with a custom script, but the resulting SOP outputs must still provide equivalent processed skin and ML joint data.

Reviewer 2: Would it make sense to add a screenshot here of AdnMLDataProcessing showing the outputs and then add a reference to that figure?:

```suggestion
> Before running data extraction, an AdnMLDataProcessing HDA must be configured in the Houdini scene. The `sim_skin_path` and `joints_path` arguments must point to the outputs generated by this processing HDA (see figure X). The simulated skin should point to the **ADN_OUT_SKIN** output and the joints should point to the **ADN_OUT_ML_JOINTS** output.
>
> If the scene has not been prepared with an AdnMLDataProcessing HDA, the extraction will fail because the script expects the simulation skin and joint source to contain the data generated by the processing step. Advanced users may automate this preparation themselves with a custom script, but the resulting SOP outputs must still provide equivalent processed skin and ML joint data.
```

If so it would be great to follow the same names that you provide in the script example below.

Author: Done

- `inputs.csv`: Input data containing the selected joint transforms.
- `outputs.csv`: Output data containing the extracted mesh displacement data and optional muscle activation data.
- `joints.json`: Joint hierarchy data exported in the same order used by `inputs.csv`.
- `extraction_config.json`: Configuration data describing the extraction settings used for the exported dataset.

The exported joint data uses Houdini/KineFX joint names internally. The `joints.json` file stores full hierarchy paths built from the KineFX primitive hierarchy, in the same order used for `inputs.csv`.

> [!NOTE]
> The `sim_skin_path` and `joints_path` must come from an AdnMLDataProcessing HDA. The script expects the simulation skin and joint source to have been prepared by the processing step before extraction is run.

## Validation

The extraction script validates the provided data before extraction starts.

The script checks that:

- The rest skin path is valid.
- The simulation skin path is valid.
- The joint source path is valid.
- The save directory path is valid.
- The rest skin and simulation skin are not the same object.

> - The rest skin and simulation skin have matching point counts.

Reviewer 2: Optional suggestion. Maybe we can mention this also in a note the first time you introduce the `rest_skin_path` and `sim_skin_path`.

Author: Done

- The rest skin and simulation skin contain point positions.
- The joint source geometry contains valid KineFX joint transform data.
- Every requested joint name exists in the joint source geometry.
- The skip frame value is a valid integer greater than or equal to `0`.
- The stabilization frame value is a valid integer greater than or equal to `0`.
- The frame windows are valid.
- Existing dataset files are not overwritten unless `force_overwrite` is enabled.
- The scene contains a valid AdnMLDataProcessing setup, or an equivalent scripted preprocessing setup, producing the simulation skin and joint source used by `sim_skin_path` and `joints_path`.

## Errors

The `extract` function may raise the following errors:

- `ValueError`: Raised when an input path, frame setting, geometry, or joint list is invalid.
- `FileExistsError`: Raised when dataset files already exist and `force_overwrite` is disabled.
- `InterruptedError`: Raised when the extraction is interrupted and the interruption is confirmed.

## Limitations

> - The `sim_skin_path` and `joints_path` must point to valid outputs generated by an AdnMLDataProcessing HDA, or to equivalent outputs generated by a custom scripted preprocessing setup.

Reviewer 2: Is this a limitation really? I would say it's just part of the workflow in Houdini. Just a thought.

Author: You are right, removed

# Data Extraction

> 

Reviewer 3: Add "NOTE" box to notify that this feature is available only for ML bundle.

Additional extraction settings can be provided to control frame sampling, stabilization, muscle activation extraction, and overwrite behavior.

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">from adnml.scripts.houdini import data_extraction

data_extraction.extract(
    rest_skin_path="/obj/geo1/REST_SKIN",
    sim_skin_path="/obj/geo1/OUT_SKIN",
    joints_path="/obj/geo1/OUT_JOINTS",
    joint_names_list=["root", "spine", "neck", "head"],
    save_directory_path="path/to/export/folder",
    muscles_paths="/obj/geo1/ADN_OUT_MUSCLES",
    skip_frames=2,
    stabilization_frames=5,
    frame_windows=[[1, 120], [220, 360]],
    force_overwrite=False
)
</code></pre>

The optional arguments are:

- `muscles_paths`: Muscle SOP paths used to extract muscle activation data. This can be a single string, a list of strings, or a UI-style string separated by spaces, commas, tabs, or newlines. The paths can point to AdnMuscle SOPs, muscle geometry SOPs, or **ADN_OUT_** nodes.
- `skip_frames`: Number of frames to skip between recorded poses. This helps reduce redundant pose data and extract more diverse samples.

> - `stabilization_frames`: Number of endpoint stabilization cooks to run before recording data.

Reviewer 3: maybe we can use a simpler wording and anticipate a bit what it means in terms of results.

```suggestion
- `stabilization_frames`: Number of times to recook a frame before recording displacement data. This parameter  damps the motion inertia in the recorded poses.
```

Author: Done

> - `frame_windows`: List of frame windows to record. If empty, the playback range is used.

Reviewer 3: ```suggestion
- `frame_windows`: List of frame windows to record. If empty, the entire playback range is recorded.
```

Author: Done

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">frame_windows = [[1, 120], [220, 360], [500, 720]]
</code></pre>

Each entry defines a frame range using the format `[start_frame, end_frame]`.

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">from adnml.scripts.houdini import data_extraction

data_extraction.extract(
    rest_skin_path="/obj/geo1/REST_SKIN",
    sim_skin_path="/obj/geo1/OUT_SKIN",
    joints_path="/obj/geo1/OUT_JOINTS",
    joint_names_list=["root", "spine", "neck", "head"],
    save_directory_path="path/to/export/folder",
    frame_windows=[[1, 120], [220, 360], [500, 720]]
)
</code></pre>

Frame windows are useful when only specific animation ranges should contribute to the training dataset.

## Skip Frames

The `skip_frames` argument controls how many frames are skipped between recorded poses.

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">skip_frames = 2
</code></pre>

This can be used to reduce redundant pose data and extract more diverse samples.

> Lower values are recommended for fast animations, while higher values can be used for slower animations. Typical suggested values for normal animation speeds are between `2` and `5`.

Reviewer 3: ```suggestion
Lower values are recommended for fast animations, while higher values can be used for slower animations. Typical suggested values for normal animation speeds are between `2` and `5`.
The skip frames will be computed from the start of each frame window, this ensures that the starting frame of each window is always recorded in the dataset.
```

Author: Done

## Stabilization Frames

> The `stabilization_frames` argument controls how many times each recorded frame should be recooked before data is written.

Reviewer 3: ```suggestion
The `stabilization_frames` argument controls how many times each recorded frame should be recooked before displacement data is computed and written.
```

Author: Done

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">stabilization_frames = 5
</code></pre>

> This is used to stabilize the simulation dynamics before recording data. Higher values make the recorded poses lose more dynamics and converge toward a static silhouette. Well stabilized data is required for good ML deformation training.

Reviewer 3: ```suggestion
This is used to stabilize the simulation dynamics before recording data. Higher values make each of the recorded poses lose more dynamics and converge toward a static silhouette. Well stabilized data is required for good ML deformation training.
```

Author: Done

Typical suggested values for normal animation speeds are between `5` and `10`. Faster animations may require more stabilization frames.

> [!NOTE]
> Increasing `stabilization_frames` will increase the total extraction time.

## Muscle Activation Data

To extract muscle activation data, provide the `muscles_paths` argument.

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">from adnml.scripts.houdini import data_extraction

data_extraction.extract(
    rest_skin_path="/obj/geo1/REST_SKIN",
    sim_skin_path="/obj/geo1/OUT_SKIN",
    joints_path="/obj/geo1/OUT_JOINTS",
    joint_names_list=["root", "spine", "neck", "head"],
    save_directory_path="path/to/export/folder",
    muscles_paths="/obj/geo1/ADN_OUT_MUSCLES"
)
</code></pre>

The `muscles_paths` argument can be:

- A single SOP path.
- A list of SOP paths.
- A string containing multiple paths separated by spaces, commas, tabs, or newlines.

The paths can point to AdnMuscle SOPs, muscle geometry SOPs, or **ADN_OUT_** nodes.

> When muscle paths are provided, the extraction process gathers the closest muscle activation data and includes it in the exported dataset.

Reviewer 3: Let's make this more practical.

```suggestion
When muscle paths are provided, the extracted data will support training the ML model on material properties prediction for AdnSmartTissue. If this input is not provided, the extracted data will support training models for AdnMLDeformer only.
```

Author: Done

## Force Overwrite

By default, the extraction process will not overwrite existing dataset files.

If `inputs.csv` or `outputs.csv` already exist in the target export folder, the extraction will raise an error unless `force_overwrite` is enabled.

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">from adnml.scripts.houdini import data_extraction

data_extraction.extract(
    rest_skin_path="/obj/geo1/REST_SKIN",
    sim_skin_path="/obj/geo1/OUT_SKIN",
    joints_path="/obj/geo1/OUT_JOINTS",
    joint_names_list=["root", "spine", "neck", "head"],
    save_directory_path="path/to/export/folder",
    force_overwrite=True
)
</code></pre>

> [!WARNING]

> > Enabling `force_overwrite` can replace existing extraction data in the target folder.

Reviewer 3: ```suggestion
> Enabling `force_overwrite` will replace any existing extraction data in the target folder, use it with caution.
```

Author: Done

    joints_path="/obj/geo1/OUT_JOINTS",
    joint_names_list=["root", "spine", "neck", "head"],
    save_directory_path="path/to/export/folder",
    interruption_callback=confirm_interruption
)
</code></pre>

The callback must return:

- `True` to confirm cancellation.
- `False` to continue extracting.

If the extraction is interrupted, partial data may have already been written to the export folder.

> [!NOTE]
> UI progress and interruption dialogs are only used when Houdini UI is available. In batch or headless execution, UI progress and interruption dialogs are suppressed.

## TOP Work Item

The `work_item` argument is an advanced optional argument intended for TOP execution.

When provided, the extraction process can report cook percentage, custom state, and log messages to the TOP work item.

<pre><code style="white-space: pre; margin: 20px 0; padding: 10px; box-sizing: border-box;">data_extraction.extract(
    rest_skin_path="/obj/geo1/REST_SKIN",
    sim_skin_path="/obj/geo1/OUT_SKIN",
    joints_path="/obj/geo1/OUT_JOINTS",
    joint_names_list=["root", "spine", "neck", "head"],
    save_directory_path="path/to/export/folder",
    work_item=work_item
)
</code></pre>

> This argument is normally managed internally by the [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool) and does not need to be provided when running extraction manually from a regular Houdini Python context.

Reviewer 3: @lucasferryinbibo I would not expose these inputs in the docs. 

What do you think @carlosmonteagudoinbibo @josevalverdeinbibo ?

Reviewer 2: If it's not useful or meaningful for users then I would not expose it.

Author: Hiding it
